vtm-gdx-poi3d/src/org/oscim/gdx/poi3d/Poi3DLayer.java

```diff
@@ -112,14 +133,95 @@ public boolean process(MapTile tile, RenderBuckets buckets, MapElement element)
+                    if (element.isPoly() || element.isLine()) {
+                        GeometryBuffer geom = new GeometryBuffer(element);
+                        mTileClipper.clip(geom); // single points should already have been clipped
+
+                        points = geom.points;
+                        pointCount = element.getNumPoints() * 2;
+                        if (pointCount < 4)
+                            return false; // Elements may have no points after clipping
+                        // TODO lazy init?
+                        double scale = MercatorProjection.zoomLevelToScale(tile.zoomLevel);
+                        double lat = MercatorProjection.tileYToLatitudeWithScale(tile.tileY, scale);
+                        double modelDistPix = MercatorProjection.metersToPixelsWithScale(MODEL_DISTANCE, lat, scale);
+
+                        ArrayList<Float> polyPoints = new ArrayList<>(); // fixed array reasonable?
+                        float variation = (float) modelDistPix / 16f; // customizable
+
+                        if (geom.isPoly()) {
+                            Box box = new Box();
+                            box.setExtents(points, pointCount);
+
+                            double pixelsX = Tile.SIZE * tile.tileX;
+                            double pixelsY = Tile.SIZE * tile.tileY;
+                            box.xmin += (pixelsX + box.xmin) % modelDistPix;
+                            box.ymin += (pixelsY + box.ymin) % modelDistPix;
+
+                            while (box.xmin < box.xmax) {
+                                double tmpY = box.ymin;
+                                while (tmpY < box.ymax) {
+                                    float x = (float) box.xmin;
+                                    float y = (float) tmpY;
+                                    if (RANDOM_TRANSFORM) {
+                                        int hashX = getPosXHash(tile, x);
+                                        int hashY = getPosYHash(tile, y);
+                                        int hash = hashY * hashX;
+                                        x += (((hash + hashX) % 14) - 7) * variation;
+                                        y += (((hash + hashY) % 14) - 7) * variation;
+                                    }
+                                    if (GeometryUtils.pointInPoly(x, y, points, pointCount, 0)) {
+                                        polyPoints.add(x);
+                                        polyPoints.add(y);
+                                    }
+                                    tmpY += modelDistPix;
+                                }
+                                box.xmin += modelDistPix;
+                            }
+                        } else {
+                            // Place models on a line with a gap of MODEL_DISTANCE
+                            float[] p1 = new float[2]; // TODO avoid initialization
+                            float[] p2 = new float[2];
```

Comment:
Indeed those should move in class level to avoid continuous re-creation with each call.

Reply:
I made some tests and I cannot clearly confirm that object creation has influence on performance, in this case. Furthermore `process()` is not synchronized and can keep up with the alternative implementation (which needs to be synchronized).